libc/src/__support/threads/linux/rwlock.h

```diff
@@ -0,0 +1,518 @@
+    }
+  };
+
+private:
+  // Whether the RwLock is shared between processes.
+  bool is_pshared;
+  // Reader/Writer preference.
+  Preference preference;
+  // State to keep track of the RwLock.
+  cpp::Atomic<int32_t> state;
+  // writer_tid is used to keep track of the thread id of the writer. Notice
+  // that TLS address is not a good idea here since it may remains the same
+  // across forked processes.
+  cpp::Atomic<pid_t> writer_tid;
+  // Waiting queue to keep track of the pending readers and writers.
+  WaitingQueue queue;
+
+private:
+  // TODO: use cached thread id once implemented.
+  LIBC_INLINE static pid_t gettid() { return syscall_impl<pid_t>(SYS_gettid); }
+
+  LIBC_INLINE LockResult try_read_lock(State &old) {
+    while (LIBC_LIKELY(old.can_acquire_reader(preference))) {
+      cpp::optional<State> next = old.try_increase_reader_count();
+      if (!next)
+        return LockResult::Overflow;
+      if (LIBC_LIKELY(old.compare_exchange_weak_with(
+              state, *next, cpp::MemoryOrder::ACQUIRE,
+              cpp::MemoryOrder::RELAXED)))
+        return LockResult::Success;
+      // Notice that old is updated by the compare_exchange_weak_with function.
+    }
+    return LockResult::Busy;
+  }
+
+  LIBC_INLINE LockResult try_write_lock(State &old) {
+    // This while loop should terminate quickly
+    while (LIBC_LIKELY(old.can_acquire_writer(preference))) {
+      if (LIBC_LIKELY(old.compare_exchange_weak_with(
+              state, old.set_writer_bit(), cpp::MemoryOrder::ACQUIRE,
+              cpp::MemoryOrder::RELAXED))) {
+        writer_tid.store(gettid(), cpp::MemoryOrder::RELAXED);
+        return LockResult::Success;
+      }
+      // Notice that old is updated by the compare_exchange_weak_with function.
+    }
+    return LockResult::Busy;
+  }
+
+public:
+  LIBC_INLINE constexpr RwLock(Preference preference = Preference::Reader,
+                               bool is_pshared = false)
+      : is_pshared(is_pshared), preference(preference), state(0), writer_tid(0),
+        queue() {}
+
+  LIBC_INLINE LockResult try_read_lock() {
+    State old = State::load(state, cpp::MemoryOrder::RELAXED);
+    return try_read_lock(old);
+  }
+  LIBC_INLINE LockResult try_write_lock() {
+    State old = State::load(state, cpp::MemoryOrder::RELAXED);
+    return try_write_lock(old);
+  }
+
+private:
+  template <State (&SpinReload)(cpp::Atomic<int32_t> &, Preference, unsigned),
+            State (&SetPending)(cpp::Atomic<int32_t> &, cpp::MemoryOrder),
+            State (&ClearPending)(cpp::Atomic<int32_t> &, cpp::MemoryOrder),
+            FutexWordType &(WaitingQueue::Guard::*Serialization)(),
+            FutexWordType &(WaitingQueue::Guard::*PendingCount)(),
+            LockResult (RwLock::*TryLock)(State &),
+            long (WaitingQueue::*Wait)(FutexWordType,
+                                       cpp::optional<Futex::Timeout>, bool),
+            bool (State::*CanAcquire)(Preference) const>
```

Comment:
I think you can use a single `class T` and then implement these functions for a `Reader` type and a `Writer` type. Just like `std::priority_queue`'s `Compare` parameter. Or something like `std::allocator_traits` / `std::ranges::range` although I don't think we need to write so much for an internal function. Or simply write two versions. Honestly I prefer the last solution as I can't see too much value in abstracting it.

Reply:
the lock function is long enough that I don't want to repeat it.

Reply:
@QuarticCat actually, after switching to `Proxy` design, I can only save a few lines. However, I still prefer abstract the logic out as the locking procedure is of the most complicated logic and repeating such routine will make it hard to review and maintain.

Reply:
Fine. You convinced me.